### DISCUSSION TITLE
xmake-vscode调试时如何查看stl容器里的值？

### DISCUSSION THREAD

**Discussion by xiaomi2077** [xmake-io/xmake-vscode#117] [Q&A] [answered]
![image](https://user-images.githubusercontent.com/33375652/174473729-da7d5eb7-e467-4af2-90c0-63450e451102.png)
调试只能看到变量的地址，不能看到里面的值。launch.json通过指定`-enable-pretty-printing`能看到变量的值，如下图蓝色区域。
![image](https://user-images.githubusercontent.com/33375652/174473787-c668417f-3165-46ed-9946-57f3ac442f1a.png)

xmake-vscode插件不知道有没有这样的设置？

**Comment by waruqi** [comment]
你可以配置重写所有 launch.json中配置的

https://github.com/xmake-io/xmake-vscode/issues/77#issuecomment-960489322

可以把 setupCommands配置进去

**Comment by xiaomi2077** [comment, marked as answer]
> 可以把 setupCommands配置进去

没太搞懂写在哪里？
![image](https://user-images.githubusercontent.com/33375652/174477626-9878db56-a548-4952-995b-24728a8b8eec.png)

写在setting.json里好像不行，还是我写错了？

**Reply by waruqi** [reply]
你漏了 setupCommands 这个 key

**Reply by xiaomi2077** [reply]
试了好几种写法，都不行。烦请作者写个示例

**Reply by waruqi** [reply]
```json
"xmake.customDebugConfig": {
    "setupCommands": [
       {
           "text": "xxxx",
           "ignoreFailures": true
       }
    ]
}
```

**Reply by xiaomi2077** [reply]
还是不行。
![image](https://user-images.githubusercontent.com/33375652/174523935-8243c551-80e0-4d2f-bc03-f152d0d86eb2.png)
![image](https://user-images.githubusercontent.com/33375652/174524084-76eaca51-7c4f-4684-9510-f422fd5ac37a.png)

**Reply by waruqi** [reply]
那自己调下 https://github.com/xmake-io/xmake-vscode/blob/f5f4d20cbd424b056f90666311be5f1292dacf2d/src/debugger.ts#L162-L165

**Reply by xiaomi2077** [reply]
我看了下源码，好像自带的有`-enable-pretty-printing`，不用从customDebugConfig额外传入了
https://github.com/xmake-io/xmake-vscode/blob/f5f4d20cbd424b056f90666311be5f1292dacf2d/src/debugger.ts#L105-L121
171行调用的是不是vscode官方的api？我看传入的参数：debugConfig里所有的keyval对都在同一级。而launch.json里`-enable-pretty-printing`在`setupCommands`里，是2级结构，和不在同一级有关系吗？
https://github.com/xmake-io/xmake-vscode/blob/f5f4d20cbd424b056f90666311be5f1292dacf2d/src/debugger.ts#L171

**Reply by waruqi** [reply]
不知道，之前一直是这么配置的，也是从 launch.json 和其他插件那摘录下来的配置。。你可以自己改下插件 排查确认下

**Reply by xiaomi2077** [reply]
![图片](https://user-images.githubusercontent.com/33375652/174796028-531461b1-5777-4b70-8dd1-eedfe7f05f99.png)
好了，就是那里的问题。
```
debugConfig = { 
         name: `launch: ${targetName}`, 
         type: 'cppdbg', 
         request: 'launch', 
         program: targetProgram, 
         args: args, 
         stopAtEntry: true, 
         cwd: targetRunDir, 
         environment: [], 
         externalConsole: false, // @see https://github.com/xmake-io/xmake-vscode/issues/36 
         MIMode: "gdb", 
         miDebuggerPath: await this.findGdbPath(), 
         description: "Enable pretty-printing for gdb", 
         text: "-enable-pretty-printing", 
         ignoreFailures: true 
     }; 
```
改成这样，保持和launch.json同样的层级
```
debugConfig = {
                    name: `launch: ${targetName}`,
                    type: 'cppdbg',
                    request: 'launch',
                    program: targetProgram,
                    args: args,
                    stopAtEntry: true,
                    cwd: targetRunDir,
                    environment: [],
                    externalConsole: false, // @see https://github.com/xmake-io/xmake-vscode/issues/36
                    MIMode: "gdb",
                    miDebuggerPath: await this.findGdbPath(),
                    description: "Enable pretty-printing for gdb",
		    setupCommands:[
			{
			    text: "-enable-pretty-printing",
			    ignoreFailures: true
			}
		    ]
                                    };
```

**Reply by waruqi** [reply]
那来个 pr ?

**Reply by xiaomi2077** [reply]
我这边只有linux平台，其他平台没法测。要不我全改了，你那边审核测试下？

**Reply by waruqi** [reply]
可以
